Detect connection closed in ManagedHandler pooling

stephentoub opened the pull request:
We still need a real connection pooling implementation, but this change at least helps mitigate the case where the server closes the connection.  When we take a connection from the pool, we now poll to see if it's readable, and if it is (meaning either it's been shutdown or there's data available to read), we discard it as it's unusable.

cc: @geoffkizer

geoffkizer commented:
I hesitate to make this change without understanding the perf implications.

In the short term, I think it would be better to check for Connection: close on the response.  We need to do this anyway, regardless of other connection pooling issues.

Longer-term, there's probably a better solution than doing a read poll.  One solution would be to simply initiate an async read and let it sit.  If the server disconnects, the read will complete and we can handle it.  If we issue another request on the connection, then eventually the server will send a response and the read will complete then.

geoffkizer commented:
Re Connection: close, see #21940

stephentoub commented:
> check for Connection: close on the response. We need to do this anyway, regardless of other connection pooling issues.

I agree we need to do that anyway.

> I hesitate to make this change without understanding the perf implications.

I don't believe checking for Connection: close handles this case, though, so we're going to need such a thing anyway, as well.  It may be there's a better way to do it, e.g. issuing a pre-emptive initial ReceiveAsync with the task for it stored into the connection, and the poll can be removed at that time.  FWIW, the desktop implementation uses a similar poll technique.

stephentoub commented:
> issuing a pre-emptive initial ReceiveAsync with the task for it stored into the connection, and the poll can be removed at that time

Let me see what it would take to do this instead... might be fairly straightforward.

stephentoub commented:
@geoffkizer, I replaced the change with one based on ReadAsync.  I like this approach better.  Look ok?  I also added some connection pooling-related tests.

geoffkizer commented on src/System.Net.Http/src/System/Net/Http/Managed/HttpConnection.cs:
```diff
@@ -1075,7 +1083,22 @@ private Task FillAsync(CancellationToken cancellationToken)
             _readOffset = 0;
-            Task<int> t = _stream.ReadAsync(_readBuffer, 0, BufferSize, cancellationToken);
+            Task<int> t;
+            if (_initialRead != null)
```
When would this be null?

stephentoub replied:
On any call to FillAsync other than the first per request.

geoffkizer replied:
Right, sorry.

geoffkizer commented on src/System.Net.Http/src/System/Net/Http/Managed/HttpConnection.cs:
```diff
@@ -1263,6 +1287,18 @@ private void PutConnectionInPool()
+            try
+            {
+                Debug.Assert(_initialRead == null, $"Expected a previous initial read to already be consumed");
+                _initialRead = _stream.ReadAsync(_readBuffer, 0, _readBuffer.Length);
```
Seems like a comment is warranted here -- I know there's one above, but seems like there should be one here too.

stephentoub replied:
Ok

geoffkizer commented on src/System.Net.Http/tests/FunctionalTests/HttpClientHandlerTest.ConnectionPooling.cs:
```diff
@@ -0,0 +1,99 @@
+                        await server.SendAsync(new ArraySegment<byte>(Encoding.ASCII.GetBytes(responseBody)), SocketFlags.None);
+
+                        await request;
+                        server.Shutdown(SocketShutdown.Both);
```
It seems like there's a potential timing issue here.  The "initial read" may not necessarily complete immediately, and so we won't detect that the socket is closed until we try to write to it.  

This is actually a general issue with our connection pooling scheme that we're not addressing right now.

stephentoub replied:
Yes, it's a general issue.  I'm not sure how you avoid that other than supporting retries on the actual request.

geoffkizer replied:
Agreed.  There are some cases we may be able to automatically retry, but that gets dangerous too.

I don't want to solve the general issue here, but I do want to figure out how to make the test robust.

stephentoub replied:
> I do want to figure out how to make the test robust

Suggestion?  Are you just suggesting adding a retry to the test in case we happen to hit this corner?

geoffkizer replied:
Honestly, I'm not sure.  Another option would be a brief delay to try to make it more likely that the client detects the disconnect.  

Retry on disconnect seems like it defeats the purpose of the test.

stephentoub replied:
> Retry on disconnect seems like it defeats the purpose of the test.

Really? The connection will never be reusable, so there really can't be false negatives, and retry would handle false positives due to the aforementioned race condition. We could add a small delay and a few retries and that should hopefully take care of it.

geoffkizer commented on src/System.Net.Http/src/System/Net/Http/Managed/HttpConnectionPool.cs:
```diff
@@ -21,9 +21,16 @@ public HttpConnectionPool()
             {
+                if (connection.InitialReadCompleted)
+                {
+                    // We got a connection, but it was either already closed by the server or the
```
It would be nice if we took the connection out of the pool as soon as the read completed.  That said, I don't think we necessarily need to do this as part of this PR, but it would be nice to add a comment that we should revisit.

stephentoub replied:
I agree it would be nice.  It's also going to add some expense, as we'd need to create a continuation we otherwise wouldn't need to.  My expectation is that in cases where we care about performance/throughput, the majority of connections will get reused and very few will hit this (once the connection:close change goes in), in which case it should be relatively rare that connections we take from the pool have abruptly closed, and it's not worth paying extra for such cases.

geoffkizer replied:
Makes sense.

We need some sort of client timeout for idle connections anyway.  As part of that logic we can potentially check for server disconnect as well.  Won't catch it right away, but that seems fine.

stephentoub replied:
Yup

geoffkizer commented on src/System.Net.Http/src/System/Net/Http/Managed/HttpConnection.cs:
```diff
@@ -554,6 +555,11 @@ public override Task FinishAsync(CancellationToken cancellationToken)
+            // Issue a pre-emptive read on the stream, to allow us to determine
+            // whether the pooled connection has been closed and whether any
+            // erroneous data has been written to it.
+            _initialRead = _stream.ReadAsync(_readBuffer, 0, _readBuffer.Length);
```
This isn't strictly necessary, is it?  If so, the comment is misleading.  It certainly does hurt to do this, but is it the right thing?

stephentoub replied:
> This isn't strictly necessary, is it?

If we're going to always immediately use the connection when it's created afresh, then no, it shouldn't be necessary.  I can remove it.

geoffkizer replied:
I meant "does not hurt" above, sorry

geoffkizer commented on src/System.Net.Http/src/System/Net/Http/Managed/HttpConnection.cs:
```diff
@@ -572,6 +578,8 @@ public void Dispose()
             }
         }
 
+        public bool InitialReadCompleted => _initialRead != null && _initialRead.IsCompleted;
```
When would _initialRead be null here?

stephentoub replied:
Based on current usage, it shouldn't be.  I could change it to an assert if you'd prefer.

geoffkizer replied:
An assert seems better to me.

geoffkizer commented on src/System.Net.Http/src/System/Net/Http/Managed/HttpConnection.cs:
```diff
@@ -37,6 +37,7 @@ internal sealed class HttpConnection : IDisposable
         private readonly byte[] _writeBuffer;
         private int _writeOffset;
 
+        private Task<int> _initialRead;
```
Can we call this something like _readAhead?  That seems more intuitive to me.  The point here isn't really so much that it's the initial read, as that it's a read issued in advance of the next request.

stephentoub replied:
Sure
